File: android-lib/src/main/java/com/kinvey/android/cache/TableNameManager.java
package com.kinvey.android.cache;


import java.util.UUID;

import io.realm.DynamicRealm;
import io.realm.DynamicRealmObject;
import io.realm.FieldAttribute;
import io.realm.RealmObjectSchema;

/**
 * Created by yuliya on 10/12/17.
 */
public class TableNameManager {

    private static final String COLLECTION_NAME = "_tableManager";
    private static final String ORIGINAL_NAME_FIELD = "originalName";
    private static final String SHORT_NAME_FIELD = "optimizedName";

    private static void initTable(DynamicRealm realm) {
        if (realm.getSchema().get(COLLECTION_NAME) == null) {
            RealmObjectSchema realmObjectSchema = realm.getSchema().create(COLLECTION_NAME);
            realmObjectSchema.addField(SHORT_NAME_FIELD, String.class, FieldAttribute.PRIMARY_KEY, FieldAttribute.REQUIRED);
            realmObjectSchema.addField(ORIGINAL_NAME_FIELD, String.class, FieldAttribute.REQUIRED);
        }
    }

    static String createShortName(String originalName, DynamicRealm realm) {
        initTable(realm);
        String shortName = UUID.randomUUID().toString();
        DynamicRealmObject object = realm.createObject(COLLECTION_NAME, shortName);
        object.set(ORIGINAL_NAME_FIELD, originalName);
        return shortName;
    }

    static String getShortName(String originalName, DynamicRealm realm) {
        initTable(realm);
        DynamicRealmObject realmObject = realm.where(COLLECTION_NAME).equalTo(ORIGINAL_NAME_FIELD, originalName).findFirst();
        return realmObject != null ? realmObject.getString(SHORT_NAME_FIELD) : null;
    }

}

File: android-lib/src/main/java/com/kinvey/android/cache/TableNameManager.kt
package com.kinvey.android.cache

import io.realm.DynamicRealm
import io.realm.DynamicRealmObject
import io.realm.FieldAttribute
import io.realm.RealmObjectSchema
import java.util.*

/**
 * Created by yuliya on 10/12/17.
 */
object TableNameManager {

    private const val COLLECTION_NAME = "_tableManager"
    private const val ORIGINAL_NAME_FIELD = "originalName"
    private const val SHORT_NAME_FIELD = "optimizedName"

    private fun initTable(realm: DynamicRealm) {
        if (realm.schema.get(COLLECTION_NAME) == null) {
            val realmObjectSchema: RealmObjectSchema = realm.schema.create(COLLECTION_NAME)
            realmObjectSchema.addField(SHORT_NAME_FIELD, String::class.java, FieldAttribute.PRIMARY_KEY, FieldAttribute.REQUIRED)
            realmObjectSchema.addField(ORIGINAL_NAME_FIELD, String::class.java, FieldAttribute.REQUIRED)
        }
    }

    fun createShortName(originalName: String?, realm: DynamicRealm): String {
        initTable(realm)
        val shortName = UUID.randomUUID().toString()
        val `object`: DynamicRealmObject = realm.createObject(COLLECTION_NAME, shortName)
        `object`.set(ORIGINAL_NAME_FIELD, originalName)
        return shortName
    }

    fun removeShortName(originalName: String?, realm: DynamicRealm) {
        initTable(realm)
        realm.where(COLLECTION_NAME).equalTo(SHORT_NAME_FIELD, originalName).findFirst()!!.deleteFromRealm()
    }

    fun getShortName(originalName: String?, realm: DynamicRealm): String {
        initTable(realm)
        val realmObject = realm.where(COLLECTION_NAME).equalTo(ORIGINAL_NAME_FIELD, originalName).findFirst()
        return realmObject?.getString(SHORT_NAME_FIELD) ?: ""
    }

    fun getOriginalName(shortName: String?, realm: DynamicRealm): String? {
        initTable(realm)
        val realmObject = realm.where(COLLECTION_NAME).equalTo(SHORT_NAME_FIELD, shortName).findFirst()
        return realmObject?.getString(ORIGINAL_NAME_FIELD)
    }
}